lib/Utils/Layout/Utils.cpp

```diff
@@ -215,14 +240,93 @@ presburger::IntegerRelation getPerRowLayoutRelation(RankedTensorType matrixType,
   auto slotMod = addModConstraint(result, slotCoeffs, paddedCols);
 
   // slotMod - col = 0
-  SmallVector<int64_t> eqConstraint(result.getNumCols(), 0);
-  eqConstraint[slotMod] = 1;
-  eqConstraint[result.getVarKindOffset(VarKind::Domain) + 1] = -1;
-  result.addEquality(eqConstraint);
+  addConstraint(
+      result,
+      {{slotMod, 1}, {result.getVarKindOffset(VarKind::Domain) + 1, -1}},
+      /*equality=*/true);
 
   return result;
 }
 
+presburger::IntegerRelation get2dConvFilterRelation(RankedTensorType filterType,
+                                                    RankedTensorType dataType,
+                                                    int64_t padding) {
+  auto domainSize = filterType.getRank();
+  assert(domainSize == 2 && "expected 2-D filter matrix");
+
+  IntegerRelation result(PresburgerSpace::getRelationSpace(
+      domainSize, /*numRange=*/2, /*numSymbol=*/0, /*numLocals=*/2));
+
+  // These are the indices that represent the position of the filter as it
+  // slides over the data.
+  auto dataRow = result.getVarKindOffset(VarKind::Local);
```

Comment:
Comment with the definition of the abbreviated names used below (idr = "index of data row"), and similar for the rest.